forte/data/extractor/attribute_extractor.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Dict, Any, Union, Iterable
+from collections import abc
+from ft.onto.base_ontology import Entry, Annotation
+from forte.common.configuration import Config
+from forte.data.data_pack import DataPack
+from forte.data.converter.feature import Feature
+from forte.data.extractor.base_extractor import BaseExtractor
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "AttributeExtractor"
+]
+
+
+class AttributeExtractor(BaseExtractor):
+    r"""AttributeExtractor extracts feature from the attribute of entry.
+
+    Args:
+        config: An instance of `Dict` or
+            :class:`forte.common.configuration.Config`
+
+            attribute: str. Required. The attribute name of the
+                entry from which features will be extracted. For
+                example, "text" attribute of Token.
+    """
+    def __init__(self, config: Union[Dict, Config]):
+        super().__init__(config)
+
+        if "attribute" not in self.config:
+            raise AttributeError("attribute needs to be specified in "
+                                "the configuration of an AttributeExtractor.")
+
+    @property
+    def state(self) -> Dict:
+        state = super().state
+        state.update({
+            "attribute": self.config.attribute
+        })
+        return state
+
+    @classmethod
+    def from_state(cls, state: Dict) \
+        -> Union["BaseExtractor", "AttributeExtractor"]:
```

Comment:
Inheriting class method is a little bit confuse. I am not sure whether there is a better way to write this.
At the same time, `Union["BaseExtractor", "AttributeExtractor"]` is my workaround for mypy type checking.

Reply:
I see. This may be tricky for users to development. When I look at this, it seems this can be implemented in the `BaseExtractor` because most code here looks generic?

Reply:
I think we need to overwrite this function for each extractor. Because there will be some special field that you need to load for different extractor and you cannot assume they exists in the BaseExtractor.

Reply:
The implementation you provided here actually is class agnostic right? Which part of this is class dependent?

Reply:
I think we might be talking different things here. Here is how I am planning to implement this function.

1. Any subclass from BaseExtractor will first call `BaseExtractor.from_state` to get a BaseExtractor object. 
2. Then this function tries to convert the BaseExtractor into an object of its own class.
     a. adding more fields
     b. change class by `obj.__class__`, but mypy checking cannot recognize this, so I have to write `Union["BaseExtractor", "AttributeExtractor"]` as return type.
3. in the future, if there are more than two layers of inheritance, let's say A(parent) -> B(subclass) -> C(subsbuclass). Inside the C.from_state, B.from_state should be first called and a B object is created. Then convert that B obj to a C obj.

Reply:
So I think there are two parts in the state:

1. The `Config` object. For this, we should be able to dump and load with default python serialization? Just up the states from `Config` to a key `config`. So you don't need to write the key for each one. In this case, all subclasses don't have to do the same thing again.
2. The `Vocab` object, you can also do the same? 

At least for these objects, we shouldn't repeat the implementation. 

In fact, see my suggestion above of using "Resources", where we can put vocab inside, I actually suggest we put all other states there too, so we don't have to worry about states per class.

So the pseudo code in the base class is like:

```python
def __state__(self):
   return {
       'config': self.config.to_dict(),
       'vocab': self.vocab.state(),  # may remove if we use resource
   }

def __getstate__(self, state):
    self.config = Config.parse(state['config'])
    self.vocab = self.vocab.from_state(state['vocab']) # may remove if we use resource
```

By the way, why do we need to use classmethod here? Python's default pickle functions are these: https://docs.python.org/3/library/pickle.html#object.__getstate__